Portability fixes for SchemaBuilder.

nkovacs opened the pull request:
Removed length from integer and boolean types. This is not supported on PostgreSQL,
and is only used for display in MySQL, not input validation.

Removed length from text and binary type. This is not supported on most RDBMS's and ignored by others.

Changed binary mapping to `varbinary(max)` on MSSQL instead of `binary(1)` (which requires a length and is fixed length) to better match others.

Changed timestamp mapping to datetime on MSSQL. Timestamp is a unique number on MSSQL, not a timestamp.

Removed precision and scale parameters from floating point types. This is not supported by PostgreSQL, among others.

Renamed timestamp length parameter to precision. The precision parameter is not supported on all RDBMS's, and may throw an error. This can be fixed either by removing the precision parameter, or overriding the SchemaBuilder of those RDBMS's to ignore the precision parameter.
**This still needs to be fixed**, though  it's not a problem on MySQL and PostgresQL.

Since the release date of 2.0.6 is close, I've opted to remove these parameters. Removing them later would require a BC break. Adding them in later (though I don't think that's a good idea) would not.
I think it would be a better idea to add a `length`, `precision` etc. function, so that if you know you're using a DB that supports these, you can do:

```
    'foo' => Schema::integer()->length(10)
    'bar' => Schema::float()->precision(4)->scale(4)
```

See #9102

nkovacs committed:
Portability fixes for SchemaBuilder.

Removed length from integer and boolean types. This is not supported on PostgreSQL,
and is only used for display in MySQL, not input validation.

Removed length from text and binary type. This is not supported on most RDBMS's and ignored by others.

Changed binary mapping to `varbinary(max)` on MSSQL instead of `binary(1)` (which requires a length and is fixed length)
to better match others.

Changed timestamp mapping to datetime on MSSQL. Timestamp is a unique number on MSSQL, not a timestamp.

Removed precision and scale parameters from floating point types. This is not supported by PostgreSQL, among others.

Renamed timestamp length parameter to precision. The precision parameter is not supported on all RDBMS's, and may
throw an error.

samdark commented on framework/db/mssql/QueryBuilder.php:
```diff
@@ -34,10 +34,10 @@ class QueryBuilder extends \yii\db\QueryBuilder
+        Schema::TYPE_TIMESTAMP => 'datetime',
         Schema::TYPE_TIME => 'time',
         Schema::TYPE_DATE => 'date',
-        Schema::TYPE_BINARY => 'binary(1)',
```
https://msdn.microsoft.com/en-us/library/ms188362.aspx

Fixed length makes more sense, doesn't it?

nkovacs replied:
TYPE_BINARY is blob elsewhere, which is variable length.

samdark replied:
OK.

cebe replied:
these changes are BC breaking, separated them into another PR and tagged it for 2.1: #9305

nkovacs replied:
Yeah, but this is totally broken.

cebe replied:
sure but if someone was using these in MSSQL we can not just break their code.

nkovacs replied:
I know. I guess MSSQL is not used that often, so it's not that critical.

nkovacs commented:
Ok, so QueryBuilder::getColumnType will remove the lengths if the value in typemap does not have a length, so `text(255) NOT NULL` is converted to `text NOT NULL`. This means the DB won't throw errors,  so there's no need to remove the parameter where some RDBMS's support it.

samdark commented:
@nkovacs yep. Can you adjust pull request accordingly?

nkovacs commented:
Ok, I'll put the length back where it makes sense.

nkovacs committed:
restore length parameter for integer types

samdark commented:
btw., changing public or protected method signature is always a backwards compatibility break. It doesn't matter if it's removing parameters or adding parameters.

nkovacs commented:
I still have to check how float works, since the typemap doesn't have a length.

> btw., changing public or protected method signature is always a backwards compatibility break. It doesn't matter if it's removing parameters or adding parameters.

Yeah, but if it has a default value, only subclasses are affected, and you've done that before in point releases. It's still a smaller break.

nkovacs commented:
Ok, so the problem with float and double is that the only typemap that includes a length is cubrid's, but cubrid only accepts a single parameter, not two. So there's no point in having precision and scale for those two.

nkovacs committed:
fix tests

nkovacs commented:
I had to comment out two tests on Cubrid, because they won't work without the float parameter.

samdark commented:
@nkovacs you mean Cubrid doesn't work w/o float parameter?

nkovacs commented:
I removed the precision and scale parameters from the float function, because only cubrid used it, but cubrid only accepts one parameter, and it might fail if you specify both (e.g. `float(4,4)`), and I think getColumnType will allow that.
Two of the tests tested that the float parameter works, i.e. that `Schema::float(16)` creates a column that's `float(16)`, not the default in typemap (`float(7)`), so those tests cannot pass now. For the other RDBMS's, these tests pass, because the parameters to `Schema::float` are ignored.

samdark commented:
Then it makes sense to remove these tests instead of commenting them out.

nkovacs commented:
Those tests also verify that `Schema::TYPE_FLOAT . '(16)'` is converted to `float(16)`, which works on cubrid because the typemap allows it.

nkovacs committed:
Test fixes.

Removed cubrid and oracle float tests that used parameters, since
Schema::float no longer supports parameters.

cebe commented:
why did you remove the precision from float and double then?

nkovacs commented:
Another possibility is to keep the precision and scale parameters, and extend CubridSchemaBuilder so that it ignores scale. Then you can add support for the parameters to other db's later.
